File: 2023/Day 4/python/day4.py
with open("input.txt", "r") as f:
    data = [line.strip() for line in f]

def calculate_total_points(cards: dict[int, dict[str, set[str]]], matches: dict[int, int]) -> int:
    return sum(2 ** (matches[i] - 1) for i in cards if matches[i] > 0)

def calculate_total_scratchcards(cards: dict[int, dict[str, set[str]]], matches: dict[int, int]) -> int:
    instances = {i: 1 for i in range(len(cards))}
    for i in range(len(cards)):
        for j in range(i + 1, min(len(cards) + 1, i + 1 + matches[i])):
            instances[j] += instances[i]
    return sum(instances.values())

cards = {i: {"winning": set(n.split("|")[0].split()[2:]),
             "mine": set(n.split("|")[1].split())}
         for i, n in enumerate(data)}

matches: dict[int, int] = {i: len(cards[i]["winning"].intersection(cards[i]["mine"])) for i in cards}

p1 = calculate_total_points(cards, matches)
print(p1)

p2 = calculate_total_scratchcards(cards, matches)
print(p2)



# My refactor of Mysty's attempt
p1_total = 0
p2_total = 0
thing = {i: 1 for i in range(len(data))}

for i, r in enumerate(data):
    numbers_s, winning_s = r.split(": ")[1].split(" | ")

    numbers = [int(n) for n in numbers_s.split() if n.isdigit()]
    winning = [int(n) for n in winning_s.split() if n.isdigit()]
    matching = [n for n in numbers if n in winning]

    # PART TWO
    for j in range(1, len(matching) + 1):
        thing[i + j] += thing[i]

    # PART ONE
    p1_total += 2 ** (len(matching) - 1) if matching else 0

p2_total = sum(thing.values())
print(f"PART ONE TOTAL: {p1_total}")
print(f"PART TWO TOTAL: {p2_total}")

File: 2023/Day 4/perl/day4.pl
#!/usr/bin/perl

use strict;
use warnings;

open my $fh, '<', 'input.txt' or die "Failed to open input.txt: $!";
my @lines = <$fh>;
close $fh;

my %cards;

foreach my $i (0 .. $#lines) {

    my $line = $lines[$i];
    $line =~ s/^Card \s* \d+ : \s* //x;
    my ($winning, $mine) = split /\s*\|\s*/, $line;
    my @winningNumbers = grep { $_ ne '' } split ' ', $winning;
    my @mineNumbers = grep { $_ ne '' } split ' ', $mine;

    $cards{$i + 1} = {
        'winning' => \@winningNumbers,
        'mine' => \@mineNumbers,
    };
}

sub calculateTotalPoints {
    my ($cards, $matches) = @_;
    my $total = 0;
    foreach my $match (values %$matches) {
        $total += 2 ** ($match - 1) if $match > 0;
    }
    return $total;
}

sub calculateTotalScratchcards {
    my ($cards, $matches) = @_;
    my @instances = (1) x (scalar keys %$cards);
    foreach my $i (sort { $a <=> $b } keys %$matches) {
        for (my $j = $i + 1; $j <= $i + $matches->{$i} && $j <= scalar keys %$cards; $j++) {
            $instances[$j] += $instances[$i];
        }
    }
    my $sum = 0;
    $sum += $_ for @instances;
    return $sum;
}

my %matches;
foreach my $i (keys %cards) {
    my %count;
    $count{$_}++ for @{$cards{$i}{winning}};
    $count{$_}++ for @{$cards{$i}{mine}};
    $matches{$i} = grep { $count{$_} == 2 } keys %count;
}

my $p1 = calculateTotalPoints(\%cards, \%matches);
print "$p1\n";

my $p2 = calculateTotalScratchcards(\%cards, \%matches);
print "$p2\n";


# Alternate way using Mysty's Python version as a base

my $p1_total = 0;
my $p2_total = 0;
my @thing = (1) x scalar @lines;

foreach my $i (0 .. $#lines) {
    my ($numbers_s, $winning_s) = split / \| /, (split /: /, $lines[$i])[1];

    my @numbers = grep { /^\d+$/ } split / /, $numbers_s;
    my @winning = grep { /^\d+$/ } split / /, $winning_s;
    my @matching = grep { my $num = $_; grep { $_ == $num } @winning } @numbers;

    # PART ONE
    my $match_count = scalar @matching;
    $p1_total += $match_count ? 2 ** ($match_count - 1) : 0;

    # PART TWO
    for my $j (1 .. $match_count) {
        if ($i + $j < scalar @lines) {
            $thing[$i + $j] += $thing[$i];
        }
    }
}

$p2_total += $_ for @thing;
print "PART ONE TOTAL: $p1_total\n";
print "PART TWO TOTAL: $p2_total\n";